crates/wit-parser/src/resolve.rs

```diff
@@ -1408,13 +1410,55 @@ impl Resolve {
         }
     }
 
-    fn include_stability(&self, stability: &Stability) -> bool {
-        match stability {
-            Stability::Stable { .. } | Stability::Unknown => true,
-            Stability::Unstable { feature, .. } => {
+    fn include_stability(&self, stability: &Stability, pkg_id: &PackageId) -> Result<bool> {
+        Ok(match stability {
+            Stability::Unknown => true,
+            // NOTE: deprecations are intentionally omitted -- an existing `@since` takes precedence over `@deprecated`
+            Stability::Stable { since, .. } => {
+                let Some(p) = self.packages.get(*pkg_id) else {
+                    // We can't check much without a package (possibly dealing with an item in an `UnresolvedPackage`),
+                    // @since version & deprecations can't be checked because there's no package version to compare to.
+                    //
+                    // Feature requirements on stabilized features are ignored in resolved packages, so we do the same here.
+                    return Ok(true);
+                };
+
+                // Use of feature gating with version specifiers inside a package that is not versioned is not allowed
+                let package_version = p.name.version.as_ref().with_context(|| format!("package [{}] contains a feature gate with a version specifier, so it must have a version", p.name))?;
+
+                // If the version on the feature gate is:
+                // - released, then we can include it
+                // - unreleased, then we must check the feature (if present)
+                ensure!(
+                    since <= package_version,
+                    "feature gate cannot reference unreleased version {since} of package [{}] (current version {package_version})",
+                    p.name
+                );
+
+                true
+            }
+            Stability::Unstable {
+                feature,
+                deprecated,
+            } => {
+                // If there's a package version and a deprecation, and the deprecation version
+                // is already in effect, do not include the gated item
+                if self
+                    .packages
+                    .get(*pkg_id)
+                    .and_then(|p| p.name.version.as_ref())
+                    .is_some_and(|pkg_version| {
+                        deprecated
+                            .as_ref()
+                            .is_some_and(|deprecation_version| deprecation_version <= pkg_version)
+                    })
+                {
+                    return Ok(false);
+                }
```

Comment:
I'm a bit surprised by this clause because I was thinking that deprecation is just a compiler warning, not affecting semantics like whether the API is included or not?

Reply:
Ah you're right! This should absolutely be only a warning for tooling to use! Fixed